crates/compiler/types/src/subs.rs

```diff
@@ -5406,3 +5411,65 @@ pub fn get_member_lambda_sets_at_region(subs: &Subs, var: Variable, target_regio
 
     internal_error!("No lambda set at region {} found", target_region);
 }
+
+/// Returns true iff the given type is inhabited by at least one value.
+fn is_inhabited(subs: &Subs, var: Variable) -> bool {
+    let mut stack = vec![var];
+    while let Some(var) = stack.pop() {
+        match subs.get_content_without_compacting(var) {
+            Content::FlexVar(_)
+            | Content::RigidVar(_)
+            | Content::FlexAbleVar(_, _)
+            | Content::RigidAbleVar(_, _)
+            // We don't need to look into recursion vars here, because if they show up in this
+            // position, they *must* belong to an inhabited type. That's because
+            //   - if the recursion var was inferred from a value, then we know there is a value of
+            //     the given type.
+            //   - if the recursion var comes from an explicit annotation, then it must be an a tag
+            //     union of the form `Rec : [ R1 Rec, R2 Rec, ..., Rn Rec ]`. However, such annotations
+            //     are determined as illegal and reported during canonicalization, because you
+            //     cannot have a tag union without a non-recursive variant.
+            | Content::RecursionVar { .. } => {}
+            Content::LambdaSet(_) => {}
+            Content::Structure(structure) => match structure {
+                FlatType::Apply(_, args) => stack.extend(subs.get_subs_slice(*args)),
+                FlatType::Func(args, _, ret) => {
+                    stack.extend(subs.get_subs_slice(*args));
+                    stack.push(*ret);
+                }
+                FlatType::Record(fields, ext) => {
+                    if let Ok(iter) = fields.unsorted_iterator(subs, *ext) {
+                        let field_vars = iter.map(|(_, field)| *field.as_inner());
+                        stack.extend(field_vars)
+                    }
+                }
+                FlatType::TagUnion(tags, ext) | FlatType::RecursiveTagUnion(_, tags, ext) => {
+                    let mut is_uninhabited = true;
+                    // If any tag is inhabited, the union is inhabited!
```

Comment:
So I totally don't think it matters if we support this or not, but technically...

```haskell
Absurd : [GoodLuckInstantiatingThis Absurd]
```

...but like I said, I think it's fine if we don't support this 😛

Reply:
That's an error during canonicalization!

Reply:
See the comment about recursive types above. It's safe to do this check here because such types will never be reached at this point

Reply:
```
» Absurd : [GoodLuckInstantiatingThis Absurd]
…
… 1

── CYCLIC ALIAS ────────────────────────────────────────────────────────────────

The Absurd alias is self-recursive in an invalid way:

4│      Absurd : [GoodLuckInstantiatingThis Absurd]
        ^^^^^^

Recursion in aliases is only allowed if recursion happens behind a
tagged union, at least one variant of which is not recursive.
```